src/api/java/com/minecolonies/api/util/BlockPosUtil.java

```diff
@@ -711,45 +716,57 @@ public static Direction getXZFacing(final BlockPos pos, final BlockPos neighbor)
      */
     public static ITextComponent calcDirection(@NotNull final BlockPos building, @NotNull final BlockPos pos)
     {
-        IFormattableTextComponent compoment = null;
+        Optional<IFormattableTextComponent> component = Optional.empty();
 
-        if (pos.getZ() > building.getZ() + 1 && pos.getX() > building.getX() + 1)
+        if (pos.getZ() == building.getZ() && pos.getX() == building.getX())
         {
-            compoment = new TranslationTextComponent(DIRECTION_SOUTH);
+            if (pos.getY() > building.getY())
+            {
+                component = Optional.of(new TranslationTextComponent(DIRECTION_UP));
+            }
+            else if (pos.getY() < building.getY())
+            {
+                component = Optional.of(new TranslationTextComponent(DIRECTION_DOWN));
+            }
         }
-        else if (pos.getZ() < building.getZ() - 1)
+
+        if (pos.getZ() > building.getZ())
         {
-            compoment = new TranslationTextComponent(DIRECTION_NORTH);
+            component = Optional.of(new TranslationTextComponent(DIRECTION_SOUTH));
+        }
+        else if (pos.getZ() < building.getZ())
+        {
+            component = Optional.of(new TranslationTextComponent(DIRECTION_NORTH));
         }
 
-        if (pos.getX() > building.getX() + 1)
+        if (pos.getX() > building.getX())
         {
-            if (compoment != null)
+            if (component.isPresent())
             {
-                compoment
+                component.get()
                   .append("/")
                   .append(new TranslationTextComponent(DIRECTION_EAST));
             }
             else
             {
-                compoment = new TranslationTextComponent(DIRECTION_EAST);
+                component = Optional.of(new TranslationTextComponent(DIRECTION_EAST));
             }
         }
-        else if (pos.getX() < building.getX() - 1)
+        else if (pos.getX() < building.getX())
         {
-            if (compoment != null)
+            if (component.isPresent())
             {
-                compoment
+                component.get()
                   .append("/")
                   .append(new TranslationTextComponent(DIRECTION_WEST));
             }
             else
             {
-                compoment = new TranslationTextComponent(DIRECTION_WEST);
+                component = Optional.of(new TranslationTextComponent(DIRECTION_WEST));
             }
         }
 
-        return compoment;
+        return component.orElse(new TranslationTextComponent(DIRECTION_NONE));
```

Comment:
Oh no, don't package this through an optional. null checks are fine, no need to create wrappers everywhere

Reply:
can always component == null ? NONE : component;